### PR TITLE
Support setting the _oauth2_proxy cookie expiration time

### PR DESCRIPTION
**Pull request by ryanw666**
### What does this PR do?

Adds a setting in the che CRD, `spec.networking.auth.gateway.cookieExpire`, that controls how long the `_oauth2_proxy` cookie lasts.

This PR is a draft. I'm struggling to test it because, while I can deploy the new che-operator image to my cluster, k8s won't let me add a value for the `cookieExpire` field because it doesn't recognize the new schema. I'm not sure how to update k8s's knowledge of the che CRD schema. I'm looking for help figuring that out.

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/22460

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [Development resource are up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-development-resources)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### CODE REVIEW

**Comment by AObuchow** [comment]
Hi @ryanw666 thank you for your PR. To update the CRD's (so that k8's will understand the new `cookieExpire` field) run [`make update-dev-resources`](https://github.com/eclipse-che/che-operator#update-development-resources) and create a new commit with the changes.

**Comment by AObuchow** [comment]
@ryanw666 sorry for my earlier comment which was erroneous (I had just read the PR description).

I believe [what](https://github.com/eclipse-che/che-operator#deploy-che-operator) you're looking for is:
```
make gen-chectl-tmpl TEMPLATES=<OPERATOR_RESOURCES_PATH>
chectl server:deploy -p (k8s|minikube|microk8s|docker-desktop) --che-operator-image=<YOUR_OPERATOR_IMAGE> --templates <OPERATOR_RESOURCES_PATH>
```
Doing a `make gen-chectl-tmpl` will generate the CRD templates, and using `--templates` when deploying Che with chectl will allow you to reference those templates. Hope this helps.

File: api/v2/checluster_types.go [hunk 1]
```diff
@@ -541,6 +541,9 @@ type Gateway struct {
 	// Configuration for kube-rbac-proxy within the Che gateway pod.
 	// +optional
 	KubeRbacProxy *KubeRbacProxy `json:"kubeRbacProxy,omitempty"`
+	// How long the _oauth2_proxy cookie lasts.
+	// +optional
+	CookieExpire string `json:"cookieExpire,omitempty"`
 }
```
Comment (tolusha) [inline api/v2/checluster_types.go:547]: Thank you for your contirubution.
I believe it is better to add a dedicated structure for additional `oauth-proxy` container configuration:

```go
type Gateway struct {
...
  // +optional
  oAuthProxy *OAuthProxy `json:"oAuthProxy,omitempty"`
}
```

```go
type OAuthProxy struct {
	// Expire timeframe for cookie. If set to 0, cookie becomes a session-cookie which will expire when the browser is closed.
	// +optional
	// +kubebuilder:default:=86400
	// +kubebuilder:validation:Minimum:=0
	CookieExpireSeconds *int32 `json:"cookieExpireSeconds,omitempty"`
}
```

**Comment by ryanw666** [comment]
Thanks for the comments folks. I changed the schema to contain an intermediate `oauthProxy` structure containing the setting, and found out how to make k8s accept the field, but what I'm seeing now is that the operator seemingly reverts the change to the checluster object. I use `kubectl edit -n eclipse-che checluster`, add the `cookieExpire` field, k8s says the edit was successful, then run `kubectl edit -n eclipse-che checluster` again and the field is gone. This happens even if I redeploy the entire thing using `chectl server:delete` followed by `chectl server:deploy`, and happens even if the `cookieExpire` field is set in the initial configuration file passed to `chectl server:deploy` rather than set in a subsequent edit.

The exact command I'm using to deploy is `chectl server:deploy --platform=k8s --domain=mydomain --che-operator-cr-patch-yaml=che.yaml --skip-cert-manager --che-operator-image=myimage --templates=<OPERATOR_RESOURCSE_PATH>`

**Comment by tolusha** [comment]
Hi, @ryanw666 

1. Please, update the `OAuthProxy` structure as suggested. It is better to have timeout in seconds.
```go
type OAuthProxy struct {
	// Expire timeframe for cookie. If set to 0, cookie becomes a session-cookie which will expire when the browser is closed.
	// +optional
	// +kubebuilder:default:=86400
	// +kubebuilder:validation:Minimum:=0
	CookieExpireSeconds *int32 `json:"cookieExpireSeconds,omitempty"`
}
```

2.  Run `make update-dev-resources` to update all needed resources
3. Run `make test` to run test locally
4. No it is possible to test the solution, see https://github.com/eclipse-che/che-operator/pull/1755#issuecomment-1726835391
5. Commit all changes.

**Comment by tolusha** [comment]
Also, it is needed to sign ECA, https://api.eclipse.org/git/eca/status/gh/eclipse-che/che-operator/1755

**Comment by ryanw666** [comment]
> No it is possible to test the solution, see
https://github.com/eclipse-che/che-operator/pull/1755#issuecomment-1726835391

As I stated, I have already tried this, and the changes did not persist. If this roadblock can be solved I will work on the other items here.

**Comment by tolusha** [comment]
If we are talking about tests on minikube, then it works for me:

```bash
make update-dev-resources
make docker-build docker-push IMG=docker.io/abazko/operator:test
make gen-chectl-tmpl TEMPLATES=/tmp/operator-resources
chectl server:deploy -p minikube --che-operator-image docker.io/abazko/operator:test --templates /tmp/operator-resources
kubectl patch checluster/eclipse-che --patch '{"spec": {"networking": {"auth": {"gateway": {"oauthProxy" : {"cookieExpire": "4h0m0s"}}}}}}' --type=merge -n eclipse-che
```

**Comment by ryanw666** [comment]
Ah, I did not know about the `make update-dev-resources` command. Unfortunately, this command throws this error for me when invoking `yq`:
```
Error: unknown shorthand flag: 'Y' in -Y
Usage:
  yq eval [expression] [yaml_file1]... [flags]
```

I have the latest version of `yq` installed, v4.35.1.

**Comment by ryanw666** [comment]
Also,I went to look at the Eclipse contributor license agreement, linked at https://accounts.eclipse.org/user/eca, but it seems I can't even view it without an eclipse.org account?

**Comment by tolusha** [comment]
> I have the latest version of yq installed, v4.35.1.

I see, let me come up with something

> Also,I went to look at the Eclipse contributor license agreement, linked at https://accounts.eclipse.org/user/eca, but it seems I can't even view it without an eclipse.org account?

Probably yes, to be honest I don't remember

**Comment by ryanw666** [comment]
Thank you. Unfortunately at the `chectl server:deploy` step I receive this error:
```
    Error: ENOENT: no such file or directory, open 'che-operator/templates/che-operator/kubernetes/crds/org_checluster_cr.yaml'
    Code: ENOENT
```

A bunch of suspicious output earlier in the process looks related. At the `update-dev-resources` step, I receive this output:
```
Building image eclipse/che-operator-dev...[OK]
Running make update-dev-resources
[INFO] UBI8 image registry.access.redhat.com/ubi8-minimal:8.8-1072
/bin/bash: line 1: /che-operator/bin/controller-gen: No such file or directory
make[1]: *** [Makefile:294: generate] Error 127
/bin/bash: line 1: /che-operator/bin/controller-gen: No such file or directory
/bin/bash: line 2: /che-operator/bin/kustomize: No such file or directory
/bin/bash: line 7: /che-operator/bin/kustomize: No such file or directory
[INFO] Deployments resources generated into /che-operator/deploy/deployment/kubernetes
/bin/bash: line 7: /che-operator/bin/kustomize: No such file or directory
[INFO] Deployments resources generated into /che-operator/deploy/deployment/openshift
cp: cannot stat '/che-operator/deploy/deployment/kubernetes/objects/checlusters.org.eclipse.che.CustomResourceDefinition.yaml': No such file or directory
rm: cannot remove '/che-operator/helmcharts/next/templates/eclipse-che.Namespace.yaml': No such file or directory
[INFO] HelmCharts updated /che-operator/helmcharts/next
Script execution [OK]
```
Looks like lots of errors, though the command appears to succeed and exits with status 0.

It also appears to delete a ton of git-tracked files:
```
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   api/v2/zz_generated.deepcopy.go
	modified:   deploy/deployment/kubernetes/combined.yaml
	deleted:    deploy/deployment/kubernetes/objects/che-operator-leader-election.Role.yaml
	deleted:    deploy/deployment/kubernetes/objects/che-operator-leader-election.RoleBinding.yaml
	deleted:    deploy/deployment/kubernetes/objects/che-operator-selfsigned-issuer.Issuer.yaml
	deleted:    deploy/deployment/kubernetes/objects/che-operator-service.Service.yaml
	deleted:    deploy/deployment/kubernetes/objects/che-operator-serving-cert.Certificate.yaml
	deleted:    deploy/deployment/kubernetes/objects/che-operator.ClusterRole.yaml
	deleted:    deploy/deployment/kubernetes/objects/che-operator.ClusterRoleBinding.yaml
	deleted:    deploy/deployment/kubernetes/objects/che-operator.Deployment.yaml
	deleted:    deploy/deployment/kubernetes/objects/che-operator.Role.yaml
	deleted:    deploy/deployment/kubernetes/objects/che-operator.RoleBinding.yaml
	deleted:    deploy/deployment/kubernetes/objects/che-operator.ServiceAccount.yaml
	deleted:    deploy/deployment/kubernetes/objects/checlusters.org.eclipse.che.CustomResourceDefinition.yaml
	deleted:    deploy/deployment/kubernetes/objects/eclipse-che.Namespace.yaml
	deleted:    deploy/deployment/kubernetes/objects/org.eclipse.che.MutatingWebhookConfiguration.yaml
	deleted:    deploy/deployment/kubernetes/objects/org.eclipse.che.ValidatingWebhookConfiguration.yaml
	modified:   deploy/deployment/openshift/combined.yaml
	deleted:    deploy/deployment/openshift/objects/che-operator-leader-election.Role.yaml
	deleted:    deploy/deployment/openshift/objects/che-operator-leader-election.RoleBinding.yaml
	deleted:    deploy/deployment/openshift/objects/che-operator-service.Service.yaml
	deleted:    deploy/deployment/openshift/objects/che-operator.ClusterRole.yaml
	deleted:    deploy/deployment/openshift/objects/che-operator.ClusterRoleBinding.yaml
	deleted:    deploy/deployment/openshift/objects/che-operator.Deployment.yaml
	deleted:    deploy/deployment/openshift/objects/che-operator.Role.yaml
	deleted:    deploy/deployment/openshift/objects/che-operator.RoleBinding.yaml
	deleted:    deploy/deployment/openshift/objects/che-operator.ServiceAccount.yaml
	deleted:    deploy/deployment/openshift/objects/checlusters.org.eclipse.che.CustomResourceDefinition.yaml
	deleted:    deploy/deployment/openshift/objects/eclipse-che.Namespace.yaml
	deleted:    deploy/deployment/openshift/objects/org.eclipse.che.MutatingWebhookConfiguration.yaml
	deleted:    deploy/deployment/openshift/objects/org.eclipse.che.ValidatingWebhookConfiguration.yaml
	deleted:    helmcharts/next/crds/checlusters.org.eclipse.che.CustomResourceDefinition.yaml
	deleted:    helmcharts/next/templates/che-operator-leader-election.Role.yaml
	deleted:    helmcharts/next/templates/che-operator-leader-election.RoleBinding.yaml
	deleted:    helmcharts/next/templates/che-operator-selfsigned-issuer.Issuer.yaml
	deleted:    helmcharts/next/templates/che-operator-service.Service.yaml
	deleted:    helmcharts/next/templates/che-operator-serving-cert.Certificate.yaml
	deleted:    helmcharts/next/templates/che-operator.ClusterRole.yaml
	deleted:    helmcharts/next/templates/che-operator.ClusterRoleBinding.yaml
	deleted:    helmcharts/next/templates/che-operator.Deployment.yaml
	deleted:    helmcharts/next/templates/che-operator.Role.yaml
	deleted:    helmcharts/next/templates/che-operator.RoleBinding.yaml
	deleted:    helmcharts/next/templates/che-operator.ServiceAccount.yaml
	deleted:    helmcharts/next/templates/org.eclipse.che.MutatingWebhookConfiguration.yaml
	deleted:    helmcharts/next/templates/org.eclipse.che.ValidatingWebhookConfiguration.yaml
```

I also receive this output at the `gen-chectl-tmpl` step:
```
Building image eclipse/che-operator-dev...[OK]
Running make gen-chectl-tmpl TEMPLATES=/che-operator/operator-resources
cp: cannot stat '/che-operator/deploy/deployment/kubernetes/objects/che-operator.Deployment.yaml': No such file or directory
cp: cannot stat '/che-operator/deploy/deployment/kubernetes/objects/checlusters.org.eclipse.che.CustomResourceDefinition.yaml': No such file or directory
cp: cannot stat '/che-operator/deploy/deployment/kubernetes/objects/che-operator.ServiceAccount.yaml': No such file or directory
cp: cannot stat '/che-operator/deploy/deployment/kubernetes/objects/che-operator.ClusterRole.yaml': No such file or directory
cp: cannot stat '/che-operator/deploy/deployment/kubernetes/objects/che-operator.ClusterRoleBinding.yaml': No such file or directory
cp: cannot stat '/che-operator/deploy/deployment/kubernetes/objects/che-operator.Role.yaml': No such file or directory
cp: cannot stat '/che-operator/deploy/deployment/kubernetes/objects/che-operator.RoleBinding.yaml': No such file or directory
cp: cannot stat '/che-operator/deploy/deployment/kubernetes/objects/che-operator-service.Service.yaml': No such file or directory
cp: cannot stat '/che-operator/deploy/deployment/kubernetes/objects/che-operator-serving-cert.Certificate.yaml': No such file or directory
cp: cannot stat '/che-operator/deploy/deployment/kubernetes/objects/che-operator-selfsigned-issuer.Issuer.yaml': No such file or directory
[INFO] Generated chectl templates into /che-operator/operator-resources
Script execution [OK]
```

**Comment by tolusha** [comment]
@ryanw666 
Which OS do you have?

**Comment by ryanw666** [comment]
@tolusha I am runnning on Alpine Linux edge, but I thought that shouldn't matter because this stuff is all being run inside docker.

**Comment by tolusha** [comment]
I think so, but it fails on downloading some binaries.

Could you try the following:
1. Remove `bin` directory in che-operator folder.
2. Run in the terminal `make download-controller-gen`
3. Check the output and if `bin/controller-gen` exists?

**Comment by ryanw666** [comment]
Today I tried doing those 3 steps. The output was only `[INFO] Downloading sigs.k8s.io/controller-tools/cmd/controller-gen@v0.7.0`, and `bin/controller-gen` existed afterward, but when running the `update-dev-resources` command again, I got the same message about `/che-operator/bin/controller-gen` not existing.

**Comment by tolusha** [comment]
@ryanw666 
I see. Let me create the PR for you since I can't figure out what is wrong

**Comment by tolusha** [comment]
Closed in favor of https://github.com/eclipse-che/che-operator/pull/1760

**Pull request closed without merging**